### PR TITLE
break(iot-dev): Remove JNR unixsocket dependency

### PR DESCRIPTION
**Pull request by timtay-microsoft**
This removes our dependency on JNR unixsocket and forces users to provide their own implementation of unix domain sockets. This also adds in a sample that demonstrates how to implement this new interface using JNR unixsockets.

### CODE REVIEW

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/hsm/UnixDomainSocketChannel.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+package com.microsoft.azure.sdk.iot.device.hsm;
+
+import java.io.IOException;
+
+/**
+ * The definition for a readable and writable unix domain socket channel that may be needed in module creation depending
+ * on the Edge runtime.
+ * @see <a href="https://github.com/Azure/azure-iot-sdk-java/tree/main/device/iot-device-samples/unix-domain-socket-sample">An example implementation using JNR Unixsocket</a>
+ */
+public interface UnixDomainSocketChannel
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/hsm/UnixDomainSocketChannel.java:13]: This is the new interface that we'll expect users to implement

File: device/iot-device-samples/unix-domain-socket-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/UnixDomainSocketSample.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import java.net.URISyntaxException;
+import java.nio.ByteBuffer;
+
+
+/**
+ * Sample code that demonstrates how to implement the {@link UnixDomainSocketChannel} interface using JNR Unixsocket.
+ *
+ * Read more about JNR Unixsocket here: https://github.com/jnr/jnr-unixsocket
+ */
+public class UnixDomainSocketSample
```
Comment (timtay-microsoft) [inline device/iot-device-samples/unix-domain-socket-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/UnixDomainSocketSample.java:22]: We'll use this implementation in our gates when we run Edge tests

File: device/iot-device-samples/unix-domain-socket-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/UnixDomainSocketSample.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+{
+    public static class UnixDomainSocketChannelImpl implements UnixDomainSocketChannel
```
Comment (timtay-microsoft) [inline device/iot-device-samples/unix-domain-socket-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/UnixDomainSocketSample.java:24]: We'll use this implementation in our gates when we run Edge tests

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ModuleClient.java [hunk 4]
```diff
@@ -146,33 +147,54 @@ private ModuleClient(String hostName, String deviceId, String moduleId, SasToken
 
     /**
      * Create a module client instance from your environment variables
+     * @param unixDomainSocketChannel the implementation of the {@link UnixDomainSocketChannel} interface that will be used if any
+     * unix domain socket communication is required. May be null if no unix domain socket communication is required. If
+     * this argument is null and unix domain socket communication is required, this method will through an {@link IllegalArgumentException}.
+     * To check if unix domain socket communication is required for your Edge runtime, check its "IOTEDGE_WORKLOADURI"
+     * environment variable. If it is not present, or its value is prefixed with "HTTP" or "HTTPS", then no unix domain
+     * socket communication is required, and this argument can be set to null. If its value is present and is prefixed
+     * with "unix", then unix domain socket communication will be required, and this argument must not be null.
```
Comment (abhipsaMisra) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ModuleClient.java:156]: Nice explanation!
Q - if we can create a module client without needing to specify the unix socket domain channel, do we still have a no-args constructor? It'll save users from having to pass in `null`.

File: device/iot-device-samples/unix-domain-socket-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/UnixDomainSocketSample.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        }
+    }
+
+    public static void main(String[] args)
+        throws IOException, URISyntaxException, ModuleClientException
+    {
+        UnixDomainSocketChannel unixDomainSocketChannel = new UnixDomainSocketChannelImpl();
+
+        ModuleClient moduleClient = ModuleClient.createFromEnvironment(unixDomainSocketChannel);
+        moduleClient.open();
```
Comment (abhipsaMisra) [inline device/iot-device-samples/unix-domain-socket-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/UnixDomainSocketSample.java:64]: q - is `open` an I/o operation or does the sdk wait until a message is sent to actually open the communication channel?

**Review by abhipsaMisra — APPROVED** [review]

**Review by azabbasi — APPROVED** [review]

File: device/iot-device-samples/unix-domain-socket-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/UnixDomainSocketSample.java (hunk 5 above)
Comment (timtay-microsoft) [reply device/iot-device-samples/unix-domain-socket-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/UnixDomainSocketSample.java:64]: It is an I/O operation

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ModuleClient.java (hunk 4 above)
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ModuleClient.java:156]: We can add one later if asked for it, but I'd rather users be aware that the unix domain socket communication may be necessary. I imagine users would gravitate towards the simplest constructor as their first step, and get confused when they get an error message saying "must provide a unix domain socket implementation"

**Pull request merged**